core/lib/dal/src/blocks_dal.rs

```diff
@@ -549,14 +588,139 @@ impl BlocksDal<'_, '_> {
         Ok(storage_oracle_info.and_then(DbStorageOracleInfo::into_optional_batch_oracle_info))
     }
 
+    /// The server is doing precommits based on the txs.
+    /// That means, that the last fictive l2 block will never be included into precommit txs and it's the only way how the miniblock will have no txs.
+    /// So we have to check if the last 2 rolling txs hashes for miniblocks are equal and if yes, we can set the final precommit tx id for the batch,
+    /// and that will mean we can commit the batch.
+    pub async fn set_final_precommit_operation(&mut self) -> DalResult<()> {
```

Comment:
This is too complex for dal method. Let's move most of the logic to the caller side.

Reply:
Moved out and splited to two methods